storage: short-circuit GC queue when context expires

tbg opened the pull request:
Fallout from https://github.com/cockroachdb/cockroach/issues/18199 and corresponding
testing in https://github.com/cockroachdb/cockroach/issues/15997.

When the context is expired, there is no point in shooting off another gazillion requests.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18642)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 859 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642#-KuWUWSbk58j4gNo39tn:-KuWUWSbk58j4gNo39to:bo1e9xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/14eb792aa5dcaec2a7180d781e6e813aed3c7d54/pkg/storage/gc_queue.go#L859)):*
> ```Go
> 	// succeeded in pushing a lot of transactions, the next time around we should have
> 	// less work here and manage to get to the intents, but it would be better to let
> 	// another goroutine push intents for transactions we've handled so that work that
> ```

s/push/resolve/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642#-:-KuWUaNIwo-qG4ff0FnR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: short-circuit GC queue when context expires

Fallout from https://github.com/cockroachdb/cockroach/issues/18199 and corresponding
testing in https://github.com/cockroachdb/cockroach/issues/15997.

When the context is expired, there is no point in shooting off another gazillion requests.

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 859 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642#-KuWUWSbk58j4gNo39tn:-Ku_0Hh6YYg-cRzUrEDZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14eb792aa5dcaec2a7180d781e6e813aed3c7d54/pkg/storage/gc_queue.go#L859)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/push/resolve/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/gc_queue.go, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642#-Ku_M06AC4hoEyvKK2LD:-Ku_M06AC4hoEyvKK2LE:b-7huegl) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ccf1c3b94f6a54b99f096dc2d5850fe361b80c/pkg/storage/gc_queue.go#L415)):*
> ```Go
> 			}(); err != nil {
> 				// Ignore above error, but if context is expired, no point in keeping going.
> 				if ctx.Err() != nil {
> ```

`ctx.Err` can return different types of errors, depending on whether the context timed out or was canceled. Why don't we return `ctx.Err`? Is the idea that `err` probably already is this `ctx.Err`? Same question below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642#-Ku_M06AC4hoEyvKK2LD:-Ku_R1RHUhrZ6QAOnVO9:bp32i2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ccf1c3b94f6a54b99f096dc2d5850fe361b80c/pkg/storage/gc_queue.go#L415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`ctx.Err` can return different types of errors, depending on whether the context timed out or was canceled. Why don't we return `ctx.Err`? Is the idea that `err` probably already is this `ctx.Err`? Same question below.
</blockquote></details>

Yes, and even if it isn't, I think the last error is probably more interesting than a context expired/canceled error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18642)*
<!-- Sent from Reviewable.io -->
